docs/design/018-new-http-api.md

```diff
@@ -0,0 +1,248 @@
+
+    Internally, this function will create a new client (or reuse a global one?), and will simply act as a convenience wrapper over the underlying `Client`/`Dialer`/`Transport` implementations, which will be initialized with sane default values.
+
+- Towards the end of this phase, the API should be mostly stable, based on community feedback.
+  Small changes will be inevitable, but there should be no discussion about the overall design.
+
+
+#### Phase 5: merge into k6-core, more testing
+
+At this point the extension should be relatively featureful and stable to be useful to all k6 users.
+
+**Goals**:
+
+- Merge the extension into k6 core, and make it available to k6 Cloud users.
+
+- Continue to gather and address feedback from users, thorough testing and polishing.
+
+
+#### Phase 6: deprecate `k6/http`
+
+As the final phase, we should add deprecation warnings when `k6/http` is used, and point users to the new API.
+Eventually, months down the line, we can consider replacing `k6/http` altogether with the new module.
+
+
+### Examples
+
+- `basic-get.js`:
+  ```javascript
+  import { fetch } from 'k6/x/net/http';
+
+  export default async function () {
+    // Creates a new default Client internally.
+    const response = await fetch('https://httpbin.test.k6.io/get');
+    const jsonData = await response.json();
+    console.log(jsonData);
+  }
+  ```
+
+- `basic-post.js`:
+  ```javascript
+  import { fetch } from 'k6/x/net/http';
+
+  export default async function () {
+    await fetch('https://httpbin.test.k6.io/post', {
+      method: 'POST',
+      json: { name: 'k6' }, // automatically adds 'Content-Type: application/json' header
+    });
+  }
+  ```
+
+- `basic-get-request.js`:
+  ```javascript
+  import { fetch, Request } from 'k6/x/net/http';
+
+  export default async function () {
+    const request = new Request('https://httpbin.test.k6.io/get', {
+      headers: { 'Case-Sensitive-Header': 'somevalue' },
+      // Other options that can also be passed directly to fetch()
+    });
+    const response = await fetch(request, {
+      // Some options that will merge with or override the options specified
+      // in the Request.
+    });
+    const jsonData = await response.json();
+    console.log(jsonData);
+  }
+  ```
+
+- `advanced-get.js`:
+  ```javascript
+  import { Client } from 'k6/x/net/http';
+
+  const client = new Client({
+    proxy: 'https://myproxy',
+    forceHTTP1: true,
+    dns: { ... },
+    // other transport options: h2c, TLS, etc.
+  });
+
+  export default async function () {
+    const response = await client.get('https://httpbin.test.k6.io/get');
+    const jsonData = await response.json();
+    console.log(jsonData);
+  }
+  ```
+
+- `advanced-post-streaming.js`:
+  ```javascript
+  import { File } from 'k6/x/file';
+  import { fetch } from 'k6/x/net/http';
+
+  // Will need supporting await in init context
```

Comment:
Do we have an issue with it? It sounds like a prerequisite for the 3rd phase.

Reply:
This was my attempt at coming up with a File API, and is more of a thought experiment than anything else. #2977 goes into much more detail, and these kinds of issues should be addressed there.

This is inspired by [Deno's API](https://deno.land/manual/examples/fetch_data#files-and-streams). While they do have sync versions of most APIs, it's clear that if we want to adopt streams, we'll need to defer actual reading of files to whatever process needs it. I.e. we can't read the whole file into memory here in the init context, so it's likely that this will open a file handle only, and `file.readable` will be the stream reference. Whether that can be done without `await`, or if we'll need to support this in the init context, is an open question.

Reply:
I thought I commented this ... but :( 

(top level await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#top_level_await) is a ES2022 feature that made all the module resolutions in ESM asynchrnous. Which is partly what makes them a bit more involved and is what blocked them on async/await support in goja.

As soon as ESM is merged this will work as well. While I am in practice blocked specifically on fixing basically this functionality - I do need it for the basic implementation, so ... yeah - once ESM is native it will work.

2 workarounds are possible until then:
1. we can make the already existing wrapper async - this likely will have ... strange behaviours, but we already need to fix those around #2674 
2. The workaround for before this was a thing was to make an inline async lambda and call it. This has some usability problems, but will likely be okay for a release or two, given that I expect this functionality will not be released before ESM is on the final stretch.